lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -601,24 +643,84 @@ static llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
 
 namespace {
 
+/// State that should be shared by all elaborator and materializer instances.
+struct SharedState {
+  SharedState(SymbolTable &table, unsigned seed) : table(table), rng(seed) {}
+
+  SymbolTable &table;
+  std::mt19937 rng;
+  Namespace names;
+  Internalizer internalizer;
+};
+
+/// A collection of state per RTG test.
+struct TestState {
+  /// The name of the test.
+  StringAttr name;
+
+  /// Keep track of the sequences we have randomized and materialized for this
+  /// test. We do not have to track them in the shared state because a value
+  /// defined by a `randomize_sequence` operation cannot cross from one test to
+  /// another, but it may be passed to a nested region which is why we cannot
+  /// keep track of it in the materializer itself. We could treat it just like
+  /// all other IdentityValues, but that is not necessary for sequences since
+  /// their identity is defined by the symbol reference to the 'rtg.sequence'
+  /// op once it has been randomized instead of the SSA-value defining op.
+  DenseMap<RandomizedSequenceStorage *, SequenceOp> materializedSequences;
+
+  /// The context switches registered for this test.
+  MapVector<
+      std::pair<ContextResourceAttrInterface, ContextResourceAttrInterface>,
+      SequenceStorage *>
+      contextSwitches;
+};
+
 /// Construct an SSA value from a given elaborated value.
 class Materializer {
 public:
-  Materializer(OpBuilder builder) : builder(builder) {}
+  Materializer(OpBuilder builder, TestState &testState,
+               SharedState &sharedState,
+               SmallVector<ElaboratorValue> &blockArgs)
+      : builder(builder), testState(testState), sharedState(sharedState),
+        blockArgs(blockArgs) {}
 
   /// Materialize IR representing the provided `ElaboratorValue` and return the
   /// `Value` or a null value on failure.
   Value materialize(ElaboratorValue val, Location loc,
-                    std::queue<RandomizedSequenceStorage *> &elabRequests,
                     function_ref<InFlightDiagnostic()> emitError) {
     auto iter = materializedValues.find(val);
     if (iter != materializedValues.end())
       return iter->second;
 
     LLVM_DEBUG(llvm::dbgs() << "Materializing " << val << "\n\n");
 
+    // In debug mode, track whether values with identity were already
+    // materialized before and assert in such a situation.
     return std::visit(
-        [&](auto val) { return visit(val, loc, elabRequests, emitError); },
+        [&](auto value) {
+          if constexpr (std::is_base_of_v<IdentityValue,
+                                          std::remove_pointer_t<
+                                              std::decay_t<decltype(value)>>>) {
```

Comment:
:sunglasses: That is a pretty nice trick to defer this to the compile time, since I guess `value` will have its type inferred for the different things being visited :thinking: :rocket: